Replacing hard-coded list-style-types by predefined counter-styles

vitorroriz opened the pull request:
#### e4fbd95a91f2f73d0cf4fb7221ce4b58560a4fb6
<pre>
Replacing hard-coded list-style-types by predefined counter-styles
<a href="https://bugs.webkit.org/show_bug.cgi?id=254111">https://bugs.webkit.org/show_bug.cgi?id=254111</a>
rdar://106085436

Reviewed by Tim Nguyen.

This patch makes the list-style-types
that are supported by predefined counter-styles-at-rules
at counterSytyles.css to be handled according to these
rules if counterStylesAtRules is enabled. Otherwise,
they are handled like before.

* LayoutTests/TestExpectations:
* LayoutTests/fast/lists/decimal-leading-zero-expected.txt:
* LayoutTests/fast/lists/decimal-leading-zero.html:
* Source/WebCore/css/CSSCounterStyle.cpp:
* Source/WebCore/css/CSSCounterStyle.h:
* Source/WebCore/css/CSSCounterStyleRegistry.cpp:
(WebCore::isCounterStyleUnsupportedByUserAgent):
* Source/WebCore/css/CSSCounterStyleRegistry.h:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/counterStyles.css:
(@counter-style square):
(@counter-style cjk-earthly-branch):
(@counter-style cjk-heavenly-stem):
* Source/WebCore/editing/Editor.cpp:
(WebCore::editableTextListsAtPositionInDescendingOrder):
* Source/WebCore/editing/FontAttributes.h:
* Source/WebCore/editing/cocoa/FontAttributesCocoa.mm:
(WebCore::cocoaTextListMarkerName):
(WebCore::TextList::createTextList const):
* Source/WebCore/rendering/RenderListMarker.cpp:
(WebCore::RenderListMarker::textRun const):
(WebCore::RenderListMarker::paint):
(WebCore::RenderListMarker::updateContent):
(WebCore::RenderListMarker::computePreferredLogicalWidths):
(WebCore::RenderListMarker::updateMargins):
(WebCore::RenderListMarker::relativeMarkerRect):
(WebCore::RenderListMarker::counterStyle const):
(WebCore::RenderListMarker::widthUsesMetricsOfPrimaryFont const):
* Source/WebCore/rendering/RenderListMarker.h:
* Source/WebCore/rendering/style/ListStyleType.cpp:
(WebCore::ListStyleType::isCircle const):
(WebCore::ListStyleType::isSquare const):
(WebCore::ListStyleType::isDisc const):
(WebCore::ListStyleType::isDisclosureClosed const):
* Source/WebCore/rendering/style/ListStyleType.h:
(WebCore::ListStyleType::operator== const):
(WebCore::ListStyleType::operator!= const):
(WebCore::ListStyleType::encode const):
(WebCore::ListStyleType::decode):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertListStyleType):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/261823@main">https://commits.webkit.org/261823@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6f2cfb12ef859fd05bf6aa0aa6f43a1a144036ce

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112985 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/22133 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1658 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4758 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121470 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/117092 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23491 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/13304 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/5965 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118772 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/17451 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100709 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/106075 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/99409 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/1245 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/46478 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14433 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/1287 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/15142 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/10621 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20448 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/53278 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8256 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16991 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/CSSCounterStyleRegistry.cpp:
```diff
@@ -164,4 +164,62 @@ bool CSSCounterStyleRegistry::operator==(const CSSCounterStyleRegistry& other) c
 
+bool isCounterStyleSupportedByUserAgent(CSSValueID valueID)
+{
+    // This list has to be in sync with styles defined in counterStyles.css
```
I would prefer the opposite list: `isCounterStyleUnsupportedByUserAgent`

This way you don't need to keep lists in sync.

nt1m replied:
Also, we should remove the value IDs/ListStyleType enum values for the ones that are supported from counter styles, and the related legacy code for it.

vitorroriz replied:
We can't yet. We want to be able to turn off the feature flag for now. I think when we remove the flag we can get rid of the legacy code.

nt1m commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -2822,6 +2822,39 @@ class TextEmphasisStyleWrapper final : public DiscretePropertyWrapper<TextEmphas
 };
 
+
+class DiscreteListStyleTypeWrapper : public AnimationPropertyWrapperBase {
```
this wrapper doesn't seem needed anymore

nt1m commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -21,7 +21,7 @@
  * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
  * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
- * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
```
nit: unrelated whitespace changes throughout this file

nt1m commented on Source/WebCore/rendering/RenderListMarker.cpp:
```diff
@@ -2029,4 +2004,64 @@ CSSCounterStyle* RenderListMarker::counterStyle() const
     return document().counterStyleRegistry().resolvedCounterStyle(style().listStyleStringValue()).get();
 }
 
+bool RenderListMarker::markerNeedsCustomAdjust() const
```
not a huge fan of this name

anttijk commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -3998,7 +3998,7 @@
             "codegen-properties": {
-                "custom": "All",
+                "converter": "ListStyleType",
                 "parser-function": "consumeListStyleType",
```
Please separate out at least this refactoring so we get more reviewable patches.

vitorroriz replied:
I've created https://github.com/WebKit/WebKit/pull/11437 for the refactoring, I'll rebase the current PR on top of it.

anttijk commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -496,7 +496,7 @@ class RenderStyle {
 
     const AtomString& listStyleStringValue() const { return m_rareInheritedData->listStyleStringValue; }
-    ListStyleType listStyleType() const { return static_cast<ListStyleType>(m_inheritedFlags.listStyleType); }
+    std::pair<ListStyleType, const AtomString&> listStyleType() const { return { static_cast<ListStyleType>(m_inheritedFlags.listStyleType), listStyleStringValue() }; }
```
It would better to add a named struct for this rather than using std::pair.

nt1m commented on Source/WebCore/rendering/style/RenderStyle.h:
These changes are unrelated.

nt1m commented on Source/WebCore/editing/cocoa/FontAttributesCocoa.mm:
```diff
@@ -37,40 +38,37 @@
 namespace WebCore {
 
-static NSString *cocoaTextListMarkerName(ListStyleType::Type style, bool ordered)
+static NSString *cocoaTextListMarkerName(ListStyleType::Type styleType, const String& styleIdentifier, bool ordered)
```
Can we just pass the whole `ListStyleType` struct instead of 2 arguments?

vitorroriz replied:
I've start doing that but then we have to add serialization for the struct, which will be used only for Cocoa case. With this approach we reduce code with the cost of only an extra function argument.

nt1m replied:
Adding serialization seems fine to me?

nt1m commented on Source/WebCore/editing/FontAttributes.h:
```diff
@@ -30,14 +30,16 @@
 struct TextList {
-    ListStyleType::Type style { ListStyleType::Type::None };
+    ListStyleType::Type styleType { ListStyleType::Type::None };
+    String styleIdentifier;
```
Can we just store the whole ListStyleType struct?

nt1m commented on Source/WebCore/css/counterStyles.css:
```diff
@@ -263,7 +265,7 @@
 @counter-style square {
     system: cyclic;
-    symbols: \25AA;
+    symbols: \25A0;
```
Why is this different from the spec?

vitorroriz replied:
Because WebKit has been using this symbol instead for square and I didn't want to change how it looks like.

nt1m replied:
Can we move over the comment from RenderListMarker? 

`// CSS specification calls for U+25FE BLACK MEDIUM SMALL SQUARE; we have been using U+25A0 BLACK SQUARE instead for a long time.`

nt1m commented on Source/WebCore/css/counterStyles.css:
```diff
@@ -285,14 +287,14 @@
 
 @counter-style cjk-earthly-branch {
-    system: fixed;
+    system: alphabetic;
     symbols: "\5B50" "\4E11" "\5BC5" "\536F" "\8FB0" "\5DF3" "\5348" "\672A" "\7533" "\9149" "\620C" "\4EA5";
     /* 子 丑 寅 卯 辰 巳 午 未 申 酉 戌 亥 */
     suffix: "\3001";
 }
 
 @counter-style cjk-heavenly-stem {
-    system: fixed;
+    system: alphabetic;
```
Same question.

vitorroriz replied:
The specification says this system is `fixed`, however the related WPT tests expect `alphabetic`. Alphabetic might make sense, otherwise it would cover a range of only 11 values (the number of symbols) before falling back to decimal. The other main browsers behave as `alphabetic` as well. I'm opening a spec issue related to that, since it seems it used to be alphabetic and then spec was updated to fixed but nobody implemented it.

vitorroriz replied:
https://github.com/w3c/csswg-drafts/issues/8596

nt1m commented on Source/WebCore/rendering/RenderListMarker.cpp:
```diff
@@ -2025,9 +2001,69 @@ StringView RenderListMarker::textWithoutSuffix() const
+        return false;
+    auto counter = counterStyle();
+    return counter && counter->isDisclosureClosed();
 }
```
These methods should probably be in ListStyleType, rather than CounterStyle and RenderListMarker

vitorroriz replied:
That is very true, thanks!

nt1m commented on Source/WebCore/css/CSSCounterStyleRegistry.cpp:
```diff
@@ -164,4 +164,65 @@ bool CSSCounterStyleRegistry::operator==(const CSSCounterStyleRegistry& other) c
+    case CSSValueTradChineseInformal:
+    case CSSValueTradChineseFormal:
+    case CSSValueEthiopicNumeric:
+    case CSSValueCustomCounterStyle:
```
does that value really exist?

vitorroriz replied:
yes, in your renaming patch you haven't changed the value in CSSValueKeywords.in, but actually we don't need this value, but just the ListStyle::Type counter part. I'll remove it from the keywords!

nt1m commented on Source/WebCore/css/CSSCounterStyleRegistry.cpp:
```diff
@@ -164,4 +164,65 @@ bool CSSCounterStyleRegistry::operator==(const CSSCounterStyleRegistry& other) c
+    case CSSValueAmharicAbegede:
+    case CSSValueEthiopicAbegedeAmEt:
+    case CSSValueCjkEarthlyBranch:
+    case CSSValueCjkHeavenlyStem:
```
Aren't these 2 supported?

nt1m commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -4126,7 +4126,10 @@ static Vector<TextList> editableTextListsAtPositionInDescendingOrder(const Posit
         bool ordered = is<HTMLOListElement>(list);
-        textLists.uncheckedAppend({ list.renderer()->style().listStyleType().type , ordered ? downcast<HTMLOListElement>(list).start() : 1, ordered });
+        if (!list.renderer())
+            return { };
```
should this be `continue;` instead of `return`?

nt1m commented on Source/WebCore/css/counterStyles.css:
```diff
@@ -263,7 +265,8 @@
 @counter-style square {
     system: cyclic;
-    symbols: \25AA;
+    /*  CSS specification calls for U+25FE BLACK MEDIUM SMALL SQUARE; we have been using U+25A0 BLACK SQUARE instead for a long time. */
```
```suggestion
    /* CSS specification calls for U+25FE BLACK MEDIUM SMALL SQUARE; we have been using U+25A0 BLACK SQUARE instead for a long time. */
```
nit: small extra space

nt1m commented on Source/WebCore/css/counterStyles.css:
```diff
@@ -285,14 +288,14 @@
 }
 
 @counter-style cjk-earthly-branch {
```
Can you add a link above to the CSSWG issue?

nt1m commented on Source/WebCore/rendering/RenderListMarker.cpp:
```diff
@@ -1846,12 +1846,14 @@ void RenderListMarker::updateContent()
-        m_textWithSuffix = makeString(text, counter->suffix());
-        m_textWithoutSuffixLength = text.length();
-        m_textIsLeftToRightDirection = u_charDirection(text[0]) != U_RIGHT_TO_LEFT;
+        if (counter) {
```
Why do we need this new check? `counterStyle()` should always return a counter style in this case.

vitorroriz replied:
It should, but since the map is populated dynamically, if something that we aren't predicting goes wrong, I think it is more prudent to assert + avoid the nullptr deref than just asserting.

nt1m replied:
If we make `resolvedCounterStyle` return a `Ref<>`, I think it should give us this guarantee? If something goes wrong, we should fix it. We don't want to hide a nullptr bug behind a visual bug (no marker shown instead of decimal/bullets).

vitorroriz replied:
I agree we shouldn't hide it, that's why I'm asserting, but I'd say it is better to fail an assert in a debug build with an eventual problem than derefing a nullptr.

nt1m replied:
The ASSERT is fine, but I think we should remove the  `if (counter)` check.

nt1m commented on Source/WebCore/rendering/RenderListMarker.cpp:
```diff
@@ -2025,9 +2002,15 @@ StringView RenderListMarker::textWithoutSuffix() const
+    return document().counterStyleRegistry().resolvedCounterStyle(style().listStyleType().identifier);
+}
+
+bool RenderListMarker::markerNeedsWidthAdjust() const
```
```suggestion
bool RenderListMarker::widthUsesMetricsOfPrimaryFont() const
```

What do you think of this name?

vitorroriz replied:
what do you mean with the `MetricsOfPrimaryFont` bit?

nt1m replied:
We seem to be using `font.metricsOfPrimaryFont()` for these cases.

nt1m commented on Source/WebCore/rendering/RenderListMarker.cpp:
```diff
@@ -2025,9 +2002,15 @@ StringView RenderListMarker::textWithoutSuffix() const
+RefPtr<CSSCounterStyle> RenderListMarker::counterStyle() const
+{
+    return document().counterStyleRegistry().resolvedCounterStyle(style().listStyleType().identifier);
+}
```
This method should return nullptr for any `ListStyleType::Type` that isn't `CounterStyle` for accuracy.

nt1m replied:
Also, separate, but `document().counterStyleRegistry().resolvedCounterStyle` should return a `Ref<>` instead of `RefPtr<>`

nt1m commented on LayoutTests/TestExpectations:
```diff
@@ -4887,27 +4887,22 @@ imported/w3c/web-platform-tests/css/css-counter-styles/lao/css3-counter-styles-1
+# rdar://106193134
+imported/w3c/web-platform-tests/css/css-counter-styles/korean-hangul-formal/css3-counter-styles-055.html [ ImageOnlyFailure ]
+imported/w3c/web-platform-tests/css/css-counter-styles/korean-hanja-formal/css3-counter-styles-065.html [ ImageOnlyFailure ]
+imported/w3c/web-platform-tests/css/css-counter-styles/korean-hanja-informal/css3-counter-styles-060.html [ ImageOnlyFailure ]
```
Can we mark these 3 counter styles as unsupported in `isCounterStyleUnsupportedByUserAgent` instead of disabling the tests?

nt1m reviewed (approved):
r=me with comments addressed + assuming you've done A/B perf tests (Speedometer + PLT might be good sets)

nt1m commented on Source/WebCore/editing/FontAttributes.h:
```diff
@@ -30,14 +30,15 @@
 
 struct TextList {
-    ListStyleType::Type style { ListStyleType::Type::None };
+    ListStyleType styleType { ListStyleType::Type::None, ""_s };
```
Nit: Use nullString() or emptyString() here

vitorroriz committed:
Replacing hard-coded list-style-types by predefined counter-styles
https://bugs.webkit.org/show_bug.cgi?id=254111
rdar://106085436

Reviewed by Tim Nguyen.

This patch makes the list-style-types
that are supported by predefined counter-styles-at-rules
at counterSytyles.css to be handled according to these
rules if counterStylesAtRules is enabled. Otherwise,
they are handled like before.

* LayoutTests/TestExpectations:
* LayoutTests/fast/lists/decimal-leading-zero-expected.txt:
* LayoutTests/fast/lists/decimal-leading-zero.html:
* Source/WebCore/css/CSSCounterStyle.cpp:
* Source/WebCore/css/CSSCounterStyle.h:
* Source/WebCore/css/CSSCounterStyleRegistry.cpp:
(WebCore::isCounterStyleUnsupportedByUserAgent):
* Source/WebCore/css/CSSCounterStyleRegistry.h:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/counterStyles.css:
(@counter-style square):
(@counter-style cjk-earthly-branch):
(@counter-style cjk-heavenly-stem):
* Source/WebCore/editing/Editor.cpp:
(WebCore::editableTextListsAtPositionInDescendingOrder):
* Source/WebCore/editing/FontAttributes.h:
* Source/WebCore/editing/cocoa/FontAttributesCocoa.mm:
(WebCore::cocoaTextListMarkerName):
(WebCore::TextList::createTextList const):
* Source/WebCore/rendering/RenderListMarker.cpp:
(WebCore::RenderListMarker::textRun const):
(WebCore::RenderListMarker::paint):
(WebCore::RenderListMarker::updateContent):
(WebCore::RenderListMarker::computePreferredLogicalWidths):
(WebCore::RenderListMarker::updateMargins):
(WebCore::RenderListMarker::relativeMarkerRect):
(WebCore::RenderListMarker::counterStyle const):
(WebCore::RenderListMarker::widthUsesMetricsOfPrimaryFont const):
* Source/WebCore/rendering/RenderListMarker.h:
* Source/WebCore/rendering/style/ListStyleType.cpp:
(WebCore::ListStyleType::isCircle const):
(WebCore::ListStyleType::isSquare const):
(WebCore::ListStyleType::isDisc const):
(WebCore::ListStyleType::isDisclosureClosed const):
* Source/WebCore/rendering/style/ListStyleType.h:
(WebCore::ListStyleType::operator== const):
(WebCore::ListStyleType::operator!= const):
(WebCore::ListStyleType::encode const):
(WebCore::ListStyleType::decode):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertListStyleType):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/261823@main
